Major refactoring

Speierers opened the pull request:
# Major refactoring

This pull request introduces a significant refactoring of the Blender add-on, bringing substantial improvements in performance, usability, and features. The changes are designed to enhance the overall user experience, making it more efficient and enjoyable to work with Mitsuba in Blender.

This new version of the add-on relies on some changes to the Mitsuba changes, introduced in [this PR](https://github.com/mitsuba-renderer/mitsuba3/pull/1476)

### Overview video

https://github.com/user-attachments/assets/7fa6e3bd-b3b7-4df6-9d1e-15ef02f48346

### Key Enhancements

- **Codebase Restructuring**: The codebase has been completely reorganized for easier maintenance, better separation, readability, and scalability.
- **Performance Improvements**: Significant enhancements have been made to the Blender -> Mitsuba scene conversion process, including:
    - In-memory texture and mesh loading
     - Caching mechanism for converted textures
    - Efficient C++ implementation for writing to Blender's framebuffer
- **Interactive Viewport and Scene Updates**: Users can now add and edit materials, geometry, lights, and directly visualize the results using Mitsuba.
- **Mitsuba Plugin Support**: The add-on now ships with many Mitsuba plugins, including support for:
     - `Mix`, `math`, `separate_rgb`, `hue_saturation`, `map_range`, `clamp`, `rgb_to_bw`, `combine_color`, and `noise` texture shader nodes
- Texture UDIM support
- **Better Logging**: Improved logging mechanisms provide more detailed information and easier debugging.
- And more!

### 💥🔪 The sharp bits 🔪💥

- Test suite wasn't updated
- Viewport rendering doesn't work on MacOS. It segfaults when trying to run the Mitsuba LLVM within Blender. It seems that Blender ships with its own version of LLVM, which might cause some conflicts here. I did try to build Mitsuba so that it uses the Blender LLVM libraries, but this still didn't fix the crashes.
- I just realised that the "pip install mitsuba" feature of the add-on preference got lost during the refactoring. We might wanna add this back once the [Mitsuba PR](https://github.com/mitsuba-renderer/mitsuba3/pull/1476) has shipped.

Speierers committed:
Major refactoring

Speierers commented:
> One thing that concerns me is the full loss of the testing harness. Experience has shown that projects without one end up unhealthy. Ideally we would have at least basic coverage to ensure that nothing breaks catastrophically. Is that something you could still add?

Wouldn't it be enough to revive the previous test suite for this refactored codebase?

wjakob commented:
I would like to at least test rendering of some basic scenes to avoid catastrophic breakage. I am actually not sure what the previous test suite covered (@bathal1 , do you know?)

rtabbara committed:
Additional refactor changes

In general, Blender 3.6 LTS will stop being supported June 2025 so clean-up export/import to focus on v4.0+

* Remove curves export as uses currently uses unsupported shape plugin
* Add translucent BSDF export support
* Clean-up export principled BSDF
* Clean-up export of normal/bump BSDF
* Added brightness/contrast texture plugin
* Added invert_rgb texture plugin
* Added mix_rgb texture plugin
* Added color_ramp texture plugin

rtabbara commented:
Hi @dvicini ,

I've been looking into this and will be close to updating the PR with some more (minor) changes/additions soon.

bathal1 reviewed (changes requested):
Thank you very much for the PR @Speierers, this is very cool work !

Here is my review, you will find several comments/questions, it would be great if you could clarify them. 

My main concern with this PR is that it seems to heavily focus on rendering inside blender, at the expense of exporting. In particular, having python implementation of various Cycles nodes as Mitsuba plugins is great to render blender native scenes, but I am concerned that this will be misleading for users who want to export a scene to XML, where some materials may be exported without raising errors, and later fail when loading the scene in Mitsuba as some plugins aren't implemented. Some functionality should be added to make sure this doesn't happen, as exporting to XML for later use is still AFAIK the main use case for this addon.

A more minor comment: I did not see any remapping of roughness in the new material export code, was that somehow fixed? Blender usually squares the roughness value that the user inputs so some remapping is necessary to make renderings with Mitsuba match Blender.

To merge this PR, we would also need to: 
- [ ] Add the test suite back, which includes round-trip tests to make sure things are consistent
- [ ] Add support for installing mitsuba with pip, as this is the preferred way
- [ ] Merge the correesponding [PR](https://github.com/mitsuba-renderer/mitsuba3/pull/1476) on the Mitsuba 3 repo.

bathal1 committed:
Fixed inconsistent package names

bathal1 committed:
Added pip installation of Mitsuba back:
    - Added a checkbox to toggle between pip version and custom path
    - Require a restart whenever the mitsuba path is changed
    - Added messages for common errors

bathal1 commented:
I added the default pip install option back. @Speierers I'm interested to hear if you tested the reload of mitsuba when changing the custom path. In my experience it is a huge pain to point to a new build folder, and I couldn't get it to work today. I reverted to requiring the user to restart blender whenever such a change is made.

bathal1 committed:
Removed empty io folder

bathal1 committed:
Use scalar_rgb variant for I/O tasks

bathal1 committed:
Updated CI Mitsuba version logic

bathal1 committed:
Temporarily disabled CI matrix

bathal1 committed:
Allow all variants to be chosen

bathal1 committed:
Try to set a parallel variant on init

bathal1 committed:
Updated mitsuba pref init flag

bathal1 committed:
Fixed sensor test

wjakob commented:
Hi @bathal1 and @Speierers,

Leander Cretigny will work on the blender plugin this semester. To me, the most important thing seems to be to merge this PR and add proper support for testing. Do you agree with this assessment? Is there any feedback or suggestions you would have for him?

dvicini commented:
Baptiste and I had this WIP PR that updates the plugin to the latest Mitsuba version and were blocked due to some Windows crashes that we didn't have time to investigate: https://github.com/mitsuba-renderer/mitsuba-blender/pull/137

The PR also improves the testing setup and instructions a bit. Maybe it would be nice to first do an incremental update (as done in PR #137), and then adapt all these larger changes?
